### PR TITLE
remap types in dnlib.DotNet.Importer

### PR DESCRIPTION
**Pull request by dyarkovoy**
I'm writing assembly processing library backed by dnlib and suggest a simple improvement to the dnlib.DotNet.Importer for the cases when the imported entity references renamed types in the target module. For example, type A has been imported into target assembly as A_1. 
Now, when processing a reference to A, Importer needs a way to remap the reference to A_1.

### CODE REVIEW

File: src/DotNet/Importer.cs [hunk 1]
```diff
@@ -43,6 +43,13 @@ public enum ImporterOptions {
 		FixSignature = int.MinValue,
 	}
 
+	/// <summary>
+	/// Re-maps entities that were renamed in the target module
+	/// </summary>
+	public interface IImportMapper {
+		ITypeDefOrRef Map(ITypeDefOrRef source);
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: Missing XML doc comments

File: src/DotNet/Importer.cs [hunk 2]
```diff
@@ -52,6 +59,7 @@ public struct Importer {
 		readonly GenericParamContext gpContext;
 		RecursionCounter recursionCounter;
 		ImporterOptions options;
+		public IImportMapper Mapper;
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: It should be readonly and not public. Add a new ctor with an extra IImportMapper arg.

File: src/DotNet/Importer.cs [hunk 3]
```diff
@@ -613,6 +622,8 @@ public Importer(ModuleDef module, ImporterOptions options)
 		public ITypeDefOrRef Import(TypeDef type) {
 			if (type == null)
 				return null;
+			var mapped = Mapper?.Map(type);
+			if (mapped != null) return mapped;
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: nit: put the return on the next line. Same below in the other method.

File: src/DotNet/Importer.cs [hunk 4]
```diff
@@ -657,7 +668,11 @@ public Importer(ModuleDef module, ImporterOptions options)
 		/// </summary>
 		/// <param name="type">The type</param>
 		/// <returns>The imported type or <c>null</c></returns>
-		public ITypeDefOrRef Import(TypeRef type) => TryResolve(Import2(type));
+		public ITypeDefOrRef Import(TypeRef type) {
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: The TypeSpec overload should also call Map()

**Comment by dyarkovoy** [comment]
Agree on everything, have question about TypeSpec overload, doesn't it eventually call Import(ITypeDef type) or  Import(ITypeRef type) ? I see the following route: Import(TypeSpec) -> CreateClassOrValueType(ITypeDefOrRef)  -> Import(ITypeDefOrRef) -> Import(IType) -> Import(TypeDef or TypeRef), the last one is being mapped already.
Am I missing something here?

I'm 100% for the generic absract class with mappers for the members, will prepare a new PR shortly

**Comment by 0xd4d** [comment]
Yes forget about that, it's probably not very useful to return your own TypeSpec.

BTW, you don't need to create a new PR, just push to the same branch.

**Comment by dyarkovoy** [comment]
made the changes, let me know if anything else is needed to get it merged

File: src/DotNet/Importer.cs [hunk 5]
```diff
@@ -971,6 +1026,9 @@ public Importer(ModuleDef module, ImporterOptions options)
 				return field;
 			if (!recursionCounter.Increment())
 				return null;
+			var mapped = Mapper?.Map(field);
+			if (mapped != null)
+				return mapped;
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: Needs to call recursionCounter.Decrement().

File: src/DotNet/Importer.cs [hunk 6]
```diff
@@ -1000,6 +1058,9 @@ public Importer(ModuleDef module, ImporterOptions options)
 				return method;
 			if (!recursionCounter.Increment())
 				return null;
+			var mapped = Mapper?.Map(method);
+			if (mapped != null)
+				return mapped;
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: Same here

File: src/DotNet/Importer.cs [hunk 7]
```diff
@@ -1037,6 +1098,9 @@ public Importer(ModuleDef module, ImporterOptions options)
 				return null;
 			if (!recursionCounter.Increment())
 				return null;
+			var mapped = Mapper?.Map(memberRef);
+			if (mapped != null)
+				return mapped;
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: And here

File: src/DotNet/Importer.cs [hunk 8]
```diff
@@ -613,6 +660,9 @@ public Importer(ModuleDef module, ImporterOptions options)
 		public ITypeDefOrRef Import(TypeDef type) {
 			if (type == null)
 				return null;
+			var mapped = Mapper?.Map(type);
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: This code must be in Import2()

File: src/DotNet/Importer.cs [hunk 9]
```diff
@@ -657,7 +707,12 @@ public Importer(ModuleDef module, ImporterOptions options)
 		/// </summary>
 		/// <param name="type">The type</param>
 		/// <returns>The imported type or <c>null</c></returns>
-		public ITypeDefOrRef Import(TypeRef type) => TryResolve(Import2(type));
+		public ITypeDefOrRef Import(TypeRef type) {
+			var mapped = Mapper?.Map(type);
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: This code must be in Import2()

File: src/DotNet/Importer.cs [hunk 10]
```diff
@@ -43,13 +43,48 @@ public enum ImporterOptions {
...
+	}
+
 	/// <summary>
 	/// Imports <see cref="Type"/>s, <see cref="ConstructorInfo"/>s, <see cref="MethodInfo"/>s
 	/// and <see cref="FieldInfo"/>s as references
 	/// </summary>
 	public struct Importer {
 		readonly ModuleDef module;
 		readonly GenericParamContext gpContext;
+		readonly ImportMapper Mapper;
```
Comment (0xd4d) [inline src/DotNet/Importer.cs]: Should be lower case: `mapper`

**Comment by dyarkovoy** [comment]
Right, I overlooked RecusionCounter. Actually, I would refactor RecursionCounter.Increment() to return IDisposable so we can wrap recursion-prone chunks in the using(recursionCounter.Increment()) { } blocks and avoid having to call Decrement() altogether. Let me know if you like the idea and will do the needful, or you just want me to add missing Decrement()'s

**Comment by 0xd4d** [comment]
No refactoring is needed at this time.

**Comment by 0xd4d** [comment]
Unless you have something else to push, I'm ready to merge this.

**Comment by 0xd4d** [comment]
You can restore that change so they're in Import() and not the recursive Import2().

**Comment by dyarkovoy** [comment]
The problem with having it in Import2 is that mapping TypeDefs to TypeRefs introduces additional (huge!) overhead of resolving TypeRefs back to TypeDefs, as it uses very slow TypeDefFinder.Find()
My code that was taking 8 seconds previously, has now been running for 20 minutes and is like only 1% finished.
I reversed the change and you can merge now. 
I'm not sure this is the perfect fix though, it may be better to split Map into TypeDef and TypeRef versions, or use cache in TypeDefFinder, will have to further research this, but for now we can merge.

**Comment by 0xd4d** [comment]
TypeDefFinder can cache the result and it will be faster, but check the docs first since you have to disable it if you eg. rename types (but you can enable it afterwards)

Thanks for the PR, merging now.

**Pull request merged**